src/compiler/scala/tools/nsc/transform/Erasure.scala

```diff
@@ -1268,6 +1268,8 @@ abstract class Erasure extends InfoTransform
               finally tpt setType specialErasure(tree1.symbol)(tree1.symbol.tpe).resultType
             case ApplyDynamic(qual, Literal(Constant(bootstrapMethodRef: Symbol)) :: _) =>
               tree
+            case _: Apply if tree1 ne tree =>
+              transform(tree1)
```

Comment:
So this calls `preErase` recursively until it stabilizes. I'm not sure if it's a bit too clever for my taste.. There should at least be some comment.

Reply:
Mmhm. So, there are [a few](https://github.com/scala/scala/blob/v2.12.4/src/compiler/scala/tools/nsc/transform/CleanUp.scala#L87-L89) [places](https://github.com/scala/scala/blob/v2.12.4/src/compiler/scala/tools/nsc/transform/CleanUp.scala#L365-L367) [already](https://github.com/scala/scala/blob/v2.12.4/src/compiler/scala/tools/nsc/transform/Fields.scala#L726) where we recurse on a tree that's either freshly made, or a sub-tree of the tree that we're currently transforming.

My double-original idea, that didn't see the light of day, was something like
```diff
             else if (args.isEmpty && interceptedMethods(fn.symbol)) {
               if (poundPoundMethods.contains(fn.symbol)) {
+                val qualSym = qual.tpe.typeSymbol
+                val qual1 = transform(qual)
                 // This is unattractive, but without it we crash here on ().## because after
                 // erasure the ScalaRunTime.hash overload goes from Unit => Int to BoxedUnit => Int.
                 // This must be because some earlier transformation is being skipped on ##, but so
                 // far I don't know what.  For null we now define null.## == 0.
                 def staticsCall(methodName: TermName): Tree = {
-                  val newTree = gen.mkMethodCall(RuntimeStaticsModule, methodName, qual :: Nil)
+                  val newTree = gen.mkMethodCall(RuntimeStaticsModule, methodName, qual1 :: Nil)
                   global.typer.typed(newTree)
                 }

-                qual.tpe.typeSymbol match {
+                qualSym match {
                   case UnitClass | NullClass                    => LIT(0)
-                  case IntClass                                 => qual
-                  case s @ (ShortClass | ByteClass | CharClass) => numericConversion(qual, s)
-                  case BooleanClass                             => If(qual, LIT(true.##), LIT(false.##))
+                  case IntClass                                 => qual1
+                  case s @ (ShortClass | ByteClass | CharClass) => numericConversion(qual1, s)
+                  case BooleanClass                             => If(qual1, LIT(true.##), LIT(false.##))
                   case LongClass                                => staticsCall(nme.longHash)
                   case FloatClass                               => staticsCall(nme.floatHash)
                   case DoubleClass                              => staticsCall(nme.doubleHash)
```
wherein we fall somewhat in line with the other transformations, in recursing manually over a sub-tree. However, since this is buried two-method-calls-deep from the `traverse` method, there's not any way I know of to say "no, this tree's fine, I took care of it already" like we can elsewhere, so we'll wind up going over this tree again in the `super.traverse` call. I don't see it being harmful at this exact moment but it introduces a level of fragility. The other version of this, deemed ad-hoc, recursed only over `preEraseApply`, since the scope is smaller and therefore less code for me to reason about.

I tried very hard to justify the use of recursion here guarded only by `tree1 ne tree`; nowhere do we return an `Apply` wrapping the `Apply` that we just made in a way that will make it get changed and re-recurse (other than in cases like `##.##.##`). It's a bit strange, though. I'll gladly add a comment, I guess.

Reply:
I admire the efforts, and I'll try to donate some more time to see if even I can reason about one of the solutions.